File: src/Crawler/Analysis/Result/UrlAnalysisResult.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Analysis\Result;

use Crawler\Utils;

class UrlAnalysisResult
{
    /**
     * @var string[]
     */
    private array $ok = [];

    /**
     * @var string[]
     */
    private array $notice = [];

    /**
     * @var string[]
     */
    private array $warning = [];

    /**
     * @var string[]
     */
    private array $critical = [];

    /**
     * @var array[]
     */
    private array $okDetails = [];

    /**
     * @var array[]
     */
    private array $noticeDetails = [];

    /**
     * @var array[]
     */
    private array $warningDetails = [];

    /**
     * @var array[]
     */
    private array $criticalDetails = [];

    /**
     * Stats per analysis and severity
     * @var array[]
     */
    private array $statsPerAnalysis = [];

    public function addOk(string $message, string $analysisName, ?array $detail = null): void
    {
        $this->ok[] = $message;
        if ($detail) {
            $this->okDetails[$analysisName] = array_merge($this->okDetails[$analysisName] ?? [], $detail);;
        }

        $this->statsPerAnalysis[$analysisName]['ok'] = ($this->statsPerAnalysis[$analysisName]['ok'] ?? 0) + 1;
    }

    public function addNotice(string $message, string $analysisName, ?array $detail = null): void
    {
        $this->notice[] = $message;
        if ($detail) {
            $this->noticeDetails[$analysisName] = array_merge($this->noticeDetails[$analysisName] ?? [], $detail);
        }

        $this->statsPerAnalysis[$analysisName]['notice'] = ($this->statsPerAnalysis[$analysisName]['notice'] ?? 0) + 1;
    }

    public function addWarning(string $message, string $analysisName, ?array $detail = null): void
    {
        $this->warning[] = $message;
        if ($detail) {
            $this->warningDetails[$analysisName] = array_merge($this->warningDetails[$analysisName] ?? [], $detail);
        }

        $this->statsPerAnalysis[$analysisName]['warning'] = ($this->statsPerAnalysis[$analysisName]['warning'] ?? 0) + 1;
    }

    public function addCritical(string $message, string $analysisName, ?array $detail = null): void
    {
        $this->critical[] = $message;
        if ($detail) {
            $this->criticalDetails[$analysisName] = array_merge($this->criticalDetails[$analysisName] ?? [], $detail);;
        }

        $this->statsPerAnalysis[$analysisName]['critical'] = ($this->statsPerAnalysis[$analysisName]['critical'] ?? 0) + 1;
    }

    public function getStatsPerAnalysis(): array
    {
        return $this->statsPerAnalysis;
    }

    public function getOk(): array
    {
        return $this->ok;
    }

    public function getNotice(): array
    {
        return $this->notice;
    }

    public function getWarning(): array
    {
        return $this->warning;
    }

    public function getCritical(): array
    {
        return $this->critical;
    }

    public function getOkDetails(): array
    {
        return $this->okDetails;
    }

    public function getNoticeDetails(): array
    {
        return $this->noticeDetails;
    }

    public function getWarningDetails(): array
    {
        return $this->warningDetails;
    }

    public function getCriticalDetails(): array
    {
        return $this->criticalDetails;
    }

    public function getAllCount(): int
    {
        return count($this->ok) + count($this->notice) + count($this->warning) + count($this->critical);
    }

    public function getDetailsOfSeverityAndAnalysisName(string $severity, string $analysisName): array
    {
        switch ($severity) {
            case 'ok':
                return $this->okDetails[$analysisName] ?? [];
            case 'notice':
                return $this->noticeDetails[$analysisName] ?? [];
            case 'warning':
                return $this->warningDetails[$analysisName] ?? [];
            case 'critical':
                return $this->criticalDetails[$analysisName] ?? [];
            default:
                throw new \InvalidArgumentException('Unknown severity: ' . $severity);
        }
    }

    public function toIconString(string $okIcon = '✅', string $noticeIcon = 'ℹ️', string $warningIcon = '⚠', string $criticalIcon = '⛔'): string
    {
        $result = '';

        $countCritical = count($this->critical);
        $countWarning = count($this->warning);
        $countNotice = count($this->notice);
        $countOk = count($this->ok);

        if ($countCritical) {
            $result .= $countCritical . $criticalIcon . ' ';
        }
        if ($countWarning) {
            $result .= $countWarning . $warningIcon . ' ';
        }
        if ($countNotice) {
            $result .= $countNotice . $noticeIcon . ' ';
        }
        if ($countOk) {
            $result .= $countOk . $okIcon . ' ';
        }

        return trim($result);
    }

    public function toColorizedString(bool $stripWhitespaces = true): string
    {
        $result = '';

        $countCritical = count($this->critical);
        $countWarning = count($this->warning);
        $countNotice = count($this->notice);
        $countOk = count($this->ok);

        if ($countCritical) {
            $result .= Utils::getColorText(strval($countCritical), 'red', true) . ' / ';
        }
        if ($countWarning) {
            $result .= Utils::getColorText(strval($countWarning), 'magenta') . ' / ';
        }
        if ($countNotice) {
            $result .= Utils::getColorText(strval($countNotice), 'blue') . ' / ';
        }
        if ($countOk) {
            $result .= Utils::getColorText(strval($countOk), 'green') . ' / ';
        }

        return $stripWhitespaces ? str_replace(' ', '', trim($result, ' /')) : trim($result, ' /');
    }

    public function toNotColorizedString(bool $stripWhitespaces = true): string
    {
        $result = '';

        $countCritical = count($this->critical);
        $countWarning = count($this->warning);
        $countNotice = count($this->notice);
        $countOk = count($this->ok);

        if ($countCritical) {
            $result .= $countCritical . ' / ';
        }
        if ($countWarning) {
            $result .= $countWarning . ' / ';
        }
        if ($countNotice) {
            $result .= $countNotice . ' / ';
        }
        if ($countOk) {
            $result .= $countOk . ' / ';
        }

        return $stripWhitespaces ? str_replace(' ', '', trim($result, ' /')) : trim($result, ' /');
    }

    public function __toString(): string
    {
        return $this->toColorizedString();
    }

    /**
     * @param string $analysisName
     * @return array|array[]
     */
    public function getAllDetailsForAnalysis(string $analysisName): array
    {
        $result = [
            'ok' => [],
            'notice' => [],
            'warning' => [],
            'critical' => [],
        ];

        foreach ($this->criticalDetails[$analysisName] ?? [] as $detail) {
            $result['critical'][] = $detail;
        }
        foreach ($this->warningDetails[$analysisName] ?? [] as $detail) {
            $result['warning'][] = $detail;
        }
        foreach ($this->noticeDetails[$analysisName] ?? [] as $detail) {
            $result['notice'][] = $detail;
        }
        foreach ($this->okDetails[$analysisName] ?? [] as $detail) {
            $result['ok'][] = $detail;
        }

        return $result;
    }

}

File: src/analysis/result/url_analysis_result.rs
// SiteOne Crawler - UrlAnalysisResult
// (c) Jan Reges <jan.reges@siteone.cz>

use std::collections::HashMap;

use crate::utils;

#[derive(Debug, Clone, Default)]
pub struct UrlAnalysisResult {
    ok: Vec<String>,
    notice: Vec<String>,
    warning: Vec<String>,
    critical: Vec<String>,

    ok_details: HashMap<String, Vec<String>>,
    notice_details: HashMap<String, Vec<String>>,
    warning_details: HashMap<String, Vec<String>>,
    critical_details: HashMap<String, Vec<String>>,

    /// Stats per analysis and severity: analysis_name -> severity -> count
    stats_per_analysis: HashMap<String, HashMap<String, usize>>,
}

impl UrlAnalysisResult {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_ok(&mut self, message: String, analysis_name: &str, detail: Option<Vec<String>>) {
        self.ok.push(message);
        if let Some(d) = detail {
            self.ok_details.entry(analysis_name.to_string()).or_default().extend(d);
        }
        *self
            .stats_per_analysis
            .entry(analysis_name.to_string())
            .or_default()
            .entry("ok".to_string())
            .or_insert(0) += 1;
    }

    pub fn add_notice(&mut self, message: String, analysis_name: &str, detail: Option<Vec<String>>) {
        self.notice.push(message);
        if let Some(d) = detail {
            self.notice_details
                .entry(analysis_name.to_string())
                .or_default()
                .extend(d);
        }
        *self
            .stats_per_analysis
            .entry(analysis_name.to_string())
            .or_default()
            .entry("notice".to_string())
            .or_insert(0) += 1;
    }

    pub fn add_warning(&mut self, message: String, analysis_name: &str, detail: Option<Vec<String>>) {
        self.warning.push(message);
        if let Some(d) = detail {
            self.warning_details
                .entry(analysis_name.to_string())
                .or_default()
                .extend(d);
        }
        *self
            .stats_per_analysis
            .entry(analysis_name.to_string())
            .or_default()
            .entry("warning".to_string())
            .or_insert(0) += 1;
    }

    pub fn add_critical(&mut self, message: String, analysis_name: &str, detail: Option<Vec<String>>) {
        self.critical.push(message);
        if let Some(d) = detail {
            self.critical_details
                .entry(analysis_name.to_string())
                .or_default()
                .extend(d);
        }
        *self
            .stats_per_analysis
            .entry(analysis_name.to_string())
            .or_default()
            .entry("critical".to_string())
            .or_insert(0) += 1;
    }

    pub fn get_stats_per_analysis(&self) -> &HashMap<String, HashMap<String, usize>> {
        &self.stats_per_analysis
    }

    pub fn get_ok(&self) -> &[String] {
        &self.ok
    }

    pub fn get_notice(&self) -> &[String] {
        &self.notice
    }

    pub fn get_warning(&self) -> &[String] {
        &self.warning
    }

    pub fn get_critical(&self) -> &[String] {
        &self.critical
    }

    pub fn get_ok_details(&self) -> &HashMap<String, Vec<String>> {
        &self.ok_details
    }

    pub fn get_notice_details(&self) -> &HashMap<String, Vec<String>> {
        &self.notice_details
    }

    pub fn get_warning_details(&self) -> &HashMap<String, Vec<String>> {
        &self.warning_details
    }

    pub fn get_critical_details(&self) -> &HashMap<String, Vec<String>> {
        &self.critical_details
    }

    pub fn get_all_count(&self) -> usize {
        self.ok.len() + self.notice.len() + self.warning.len() + self.critical.len()
    }

    pub fn get_details_of_severity_and_analysis_name(&self, severity: &str, analysis_name: &str) -> Vec<String> {
        match severity {
            "ok" => self.ok_details.get(analysis_name).cloned().unwrap_or_default(),
            "notice" => self.notice_details.get(analysis_name).cloned().unwrap_or_default(),
            "warning" => self.warning_details.get(analysis_name).cloned().unwrap_or_default(),
            "critical" => self.critical_details.get(analysis_name).cloned().unwrap_or_default(),
            _ => Vec::new(),
        }
    }

    pub fn to_icon_string(&self) -> String {
        let mut result = String::new();

        let count_critical = self.critical.len();
        let count_warning = self.warning.len();
        let count_notice = self.notice.len();
        let count_ok = self.ok.len();

        if count_critical > 0 {
            result.push_str(&format!("{}\u{26d4} ", count_critical));
        }
        if count_warning > 0 {
            result.push_str(&format!("{}\u{26a0} ", count_warning));
        }
        if count_notice > 0 {
            result.push_str(&format!("{}\u{2139}\u{fe0f} ", count_notice));
        }
        if count_ok > 0 {
            result.push_str(&format!("{}\u{2705} ", count_ok));
        }

        result.trim().to_string()
    }

    pub fn to_colorized_string(&self, strip_whitespaces: bool) -> String {
        let mut result = String::new();

        let count_critical = self.critical.len();
        let count_warning = self.warning.len();
        let count_notice = self.notice.len();
        let count_ok = self.ok.len();

        if count_critical > 0 {
            result.push_str(&utils::get_color_text(&count_critical.to_string(), "red", true));
            result.push_str(" / ");
        }
        if count_warning > 0 {
            result.push_str(&utils::get_color_text(&count_warning.to_string(), "magenta", false));
            result.push_str(" / ");
        }
        if count_notice > 0 {
            result.push_str(&utils::get_color_text(&count_notice.to_string(), "blue", false));
            result.push_str(" / ");
        }
        if count_ok > 0 {
            result.push_str(&utils::get_color_text(&count_ok.to_string(), "green", false));
            result.push_str(" / ");
        }

        let trimmed = result.trim_end_matches(" / ").to_string();
        if strip_whitespaces {
            trimmed.replace(' ', "")
        } else {
            trimmed
        }
    }

    pub fn to_not_colorized_string(&self, strip_whitespaces: bool) -> String {
        let mut result = String::new();

        let count_critical = self.critical.len();
        let count_warning = self.warning.len();
        let count_notice = self.notice.len();
        let count_ok = self.ok.len();

        if count_critical > 0 {
            result.push_str(&format!("{} / ", count_critical));
        }
        if count_warning > 0 {
            result.push_str(&format!("{} / ", count_warning));
        }
        if count_notice > 0 {
            result.push_str(&format!("{} / ", count_notice));
        }
        if count_ok > 0 {
            result.push_str(&format!("{} / ", count_ok));
        }

        let trimmed = result.trim_end_matches(" / ").to_string();
        if strip_whitespaces {
            trimmed.replace(' ', "")
        } else {
            trimmed
        }
    }

    pub fn get_all_details_for_analysis(&self, analysis_name: &str) -> HashMap<String, Vec<String>> {
        let mut result = HashMap::new();
        result.insert(
            "ok".to_string(),
            self.ok_details.get(analysis_name).cloned().unwrap_or_default(),
        );
        result.insert(
            "notice".to_string(),
            self.notice_details.get(analysis_name).cloned().unwrap_or_default(),
        );
        result.insert(
            "warning".to_string(),
            self.warning_details.get(analysis_name).cloned().unwrap_or_default(),
        );
        result.insert(
            "critical".to_string(),
            self.critical_details.get(analysis_name).cloned().unwrap_or_default(),
        );
        result
    }
}

impl std::fmt::Display for UrlAnalysisResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.to_colorized_string(true))
    }
}
